migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/migration_impl.go

```diff
@@ -0,0 +1,327 @@
+package m181tom182
+
+import (
+	"context"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	apiTokenStore "github.com/stackrox/rox/migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/apitokenstore"
+	groupStore "github.com/stackrox/rox/migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/groupstore"
+	permissionSetStore "github.com/stackrox/rox/migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/permissionsetstore"
+	roleStore "github.com/stackrox/rox/migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/rolestore"
+	"github.com/stackrox/rox/migrator/types"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+)
+
+// Pre- and Post- migration role and permission set names
+const (
+	ScopeManagerRoleName = "Scope Manager"
+
+	deprecatedPrefix            = "[DEPRECATED] "
+	deprecatedDescriptionSuffix = ". DEPRECATED, please use \"Vulnerability Report Creator\" " +
+		"instead for vulnerability report management purposes"
+
+	scopeManagerPermissionSetID = "ffffffff-ffff-fff4-f5ff-fffffffffffb"
+)
+
+const (
+	batchSize = 500
+)
+
+var (
+	log = logging.LoggerForModule()
+
+	errNotFound = errors.New("not found")
+)
+
+func migrate(database *types.Databases) error {
+	ctx := sac.WithAllAccess(context.Background())
+
+	apiTokenStorage := apiTokenStore.New(database.PostgresDB)
+	groupStorage := groupStore.New(database.PostgresDB)
+	permissionSetStorage := permissionSetStore.New(database.PostgresDB)
+	roleStorage := roleStore.New(database.PostgresDB)
+
+	var err *multierror.Error
+	canRemoveRole := true
+	canRemovePermissionSet := true
```

Comment:
Why exactly do we have these two variables?

AFAIU, we only set `canRemoveRole == false` in case of an error during any action involving migrating the role.

If `canRemoveRole == false`, this also means `canRemovePermissionSet == false`. In the end, the error will be anyways returned and migration fails, so we could also skip these variables and simply return the error?  Not sure it'd actually help in continuing the migration if at the end the error will anyways be `!= nil`.

Reply:
`canRemoveRole == false` means that:
- A reference to the `Scope Manager` role was found in either the AuthProvider groups or in the API tokens.
- The insertion of the `[DEPRECATED] Scope Manager` failed, OR the update of the referenced role for any of the impacted groups or tokens failed.
This is indeed an error case, and leads to `canRemovePermissionSet == false`

`canRemovePermissionSet == false` means that there still is a role referencing the `Scope Manager` permission set, and that the permission set needs to be updated. This could be a valid functional case (e.g. the customer has created a role based on that permission set).

The chosen strategy for that migration is to migrate as much as possible so that in case of failure there remains as little corrective work as possible.

Reply:
I think it's a bit unfortunate that the migration code gets quite complex with all the indentation and cases of `canRemoveRole/PermissionSet` with in my opinion very little value (AFAIU if this migration fails, we'd most likely anyways need to provide a patch release for this?), but if you feel strong about having the strategy as-is for the migration, it's fine with me.